### ISSUE TITLE
feat: LaguerreFeatureProvider plugin for rangebar-py integration — entry-point adapter exposing curated RSI+regime columns for ClickHouse pre-computation, enabling SQL-native Gen800 regime-gated sweeps

### ISSUE THREAD

**Issue by terrylica** [terrylica/atr-adaptive-laguerre#1]
## Motivation

The `atr-adaptive-laguerre` package computes ATR-Adaptive Laguerre RSI — a recursive 4-stage filter cascade (`L0[i] = f(price[i], L0[i-1])`) that **cannot be expressed in SQL window functions**. This makes it the sole blocker preventing ClickHouse SQL from replicating Gen800 regime-conditioned pattern sweeps that were previously Python-only (backtesting.py).

The downstream package `rangebar-py` (terrylica/rangebar-py) is implementing a **FeatureProvider plugin system** (terrylica/opendeviationbar-py#97) that allows external packages to enrich range bars with additional columns before ClickHouse INSERT. This issue implements the `atr-adaptive-laguerre` side of that integration.

**Goal**: When a user does `pip install rangebar-py[laguerre]`, the Laguerre RSI features are automatically computed during bar construction and stored as columns in ClickHouse — no manual steps required.

## Architecture

```
rangebar-py                          atr-adaptive-laguerre
===========                          =====================

Rust core → OHLCV bars               (not involved)
     |
     v
discover_providers()  ───────────>   entry point: "rangebar.feature_providers"
     |                                    |
     v                                    v
provider.enrich(bars_df)  ──────>    LaguerreFeatureProvider.enrich()
     |                                    |
     |                               ATRAdaptiveLaguerreRSI.fit_transform_features()
     |                                    |
     v                                    v
bars_df + laguerre_* columns  <────  returns enriched DataFrame
     |
     v
store_bars_bulk() → ClickHouse INSERT (includes laguerre_* columns)
```

### 1. `LaguerreFeatureProvider` class

Create `src/atr_adaptive_laguerre/rangebar_plugin.py`:

```python
"""rangebar-py FeatureProvider plugin for ATR-Adaptive Laguerre RSI.

This module is discovered via the 'rangebar.feature_providers' entry point
when atr-adaptive-laguerre is installed alongside rangebar-py[laguerre].

It is NOT imported by default — only when rangebar-py calls discover_providers().
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

import numpy as np
import pandas as pd

from atr_adaptive_laguerre import ATRAdaptiveLaguerreRSI, ATRAdaptiveLaguerreRSIConfig

if TYPE_CHECKING:
    pass

logger = logging.getLogger(__name__)

# ---- Default config: curated from Gen800 best performer ----
# These parameters should be the single best Laguerre config from Gen800 sweep.
# Update when Gen800 final rankings are available.
DEFAULT_CONFIG = ATRAdaptiveLaguerreRSIConfig(
    atr_period=14,          # TODO: replace with Gen800 winner
    smoothing_period=3,     # TODO: replace with Gen800 winner
    level_up=0.75,          # TODO: replace with Gen800 winner
    level_down=0.10,        # TODO: replace with Gen800 winner
    adaptive_offset=0.50,   # TODO: replace with Gen800 winner
)

# Column prefix — all columns start with "laguerre_" for namespace isolation
_PREFIX = "laguerre_"

# Curated output columns (6 total — not the full 31 feature set)
_OUTPUT_COLUMNS = (
    "laguerre_rsi",
    "laguerre_regime",
    "laguerre_regime_strength",
    "laguerre_bars_in_regime",
    "laguerre_tail_risk_score",
    "laguerre_rsi_velocity",
)

# Mapping from fit_transform_features() column names to output column names
_FEATURE_MAP = {
    "rsi": "laguerre_rsi",
    "regime": "laguerre_regime",
    "regime_strength": "laguerre_regime_strength",
    "bars_in_regime": "laguerre_bars_in_regime",
    "tail_risk_score": "laguerre_tail_risk_score",
    "rsi_velocity": "laguerre_rsi_velocity",
}


class LaguerreFeatureProvider:
    """FeatureProvider plugin for rangebar-py.

    Computes ATR-Adaptive Laguerre RSI and derived features for each bar,
    using the curated config from Gen800 sweep results.

    Implements the rangebar-py FeatureProvider protocol:
    - name: str
    - version: str
    - columns: tuple[str, ...]
    - min_bars: int
    - enrich(bars_df, symbol, threshold_decimal_bps) -> pd.DataFrame
    """

    def __init__(self, config: ATRAdaptiveLaguerreRSIConfig | None = None):
        self._config = config or DEFAULT_CONFIG
        self._indicator = ATRAdaptiveLaguerreRSI(self._config)

    @property
    def name(self) -> str:
        return "laguerre"

    @property
    def version(self) -> str:
        from atr_adaptive_laguerre import __version__
        return __version__

    @property
    def columns(self) -> tuple[str, ...]:
        return _OUTPUT_COLUMNS

    @property
    def min_bars(self) -> int:
        return self._indicator.min_lookback

    def enrich(
        self,
        bars: pd.DataFrame,
        symbol: str,
        threshold_decimal_bps: int,
    ) -> pd.DataFrame:
        """Add laguerre_* columns to range bar DataFrame.

        Parameters
        ----------
        bars : pd.DataFrame
            Range bars with at minimum: open, high, low, close, volume.
            Index is either RangeIndex or DatetimeIndex.
            Rows are chronologically sorted, single symbol, single threshold.
        symbol : str
            Trading symbol (e.g., "BTCUSDT"). Not used in computation
            but available for config overrides per asset.
        threshold_decimal_bps : int
            Bar threshold. Not used in computation but available for
            config overrides per threshold.

        Returns
        -------
        pd.DataFrame
            Input DataFrame with 6 additional laguerre_* columns appended.
            First min_bars rows will have NaN values (warmup period).
        """
        if len(bars) < self.min_bars:
            logger.warning(
                "Only %d bars for %s@%d (need %d for warmup). "
                "All laguerre_* columns will be NaN.",
                len(bars), symbol, threshold_decimal_bps, self.min_bars,
            )
            for col in self.columns:
                bars[col] = np.nan
            return bars

        # Prepare OHLCV input for fit_transform_features()
        ohlcv = bars[["open", "high", "low", "close", "volume"]].copy()

        # Handle datetime: fit_transform_features() needs a 'date' column
        if isinstance(bars.index, pd.DatetimeIndex):
            ohlcv["date"] = bars.index
        elif "timestamp_ms" in bars.columns:
            ohlcv["date"] = pd.to_datetime(bars["timestamp_ms"], unit="ms")
        else:
            # Fallback: synthetic sequential dates (feature computation is timestamp-agnostic)
            ohlcv["date"] = pd.date_range("2020-01-01", periods=len(bars), freq="1s")

        # Compute full feature matrix (31 columns for single-interval)
        features = self._indicator.fit_transform_features(ohlcv)

        # Map curated subset to laguerre_* column names
        for src_col, dst_col in _FEATURE_MAP.items():
            if src_col in features.columns:
                bars[dst_col] = features[src_col].values
            else:
                logger.warning("Expected feature column '%s' not found", src_col)
                bars[dst_col] = np.nan

        return bars
```

### 2. Entry Point Registration

In `pyproject.toml`, add:

```toml
[project.entry-points."rangebar.feature_providers"]
laguerre = "atr_adaptive_laguerre.rangebar_plugin:LaguerreFeatureProvider"
```

This is the **only connection** between the two packages. rangebar-py discovers this entry point at runtime via `importlib.metadata.entry_points(group="rangebar.feature_providers")`.

### 3. Optional: Config Override via Environment

For advanced users who want a different Laguerre config than the default:

```python
import os

def _load_config_from_env() -> ATRAdaptiveLaguerreRSIConfig | None:
    """Load Laguerre config from RANGEBAR_LAGUERRE_* env vars if set."""
    atr = os.environ.get("RANGEBAR_LAGUERRE_ATR_PERIOD")
    if atr is None:
        return None
    return ATRAdaptiveLaguerreRSIConfig(
        atr_period=int(atr),
        smoothing_period=int(os.environ.get("RANGEBAR_LAGUERRE_SMOOTHING", "3")),
        level_up=float(os.environ.get("RANGEBAR_LAGUERRE_LEVEL_UP", "0.75")),
        level_down=float(os.environ.get("RANGEBAR_LAGUERRE_LEVEL_DOWN", "0.10")),
        adaptive_offset=float(os.environ.get("RANGEBAR_LAGUERRE_ADAPTIVE_OFFSET", "0.50")),
    )
```

This allows mise `[env]` to override the default without code changes.

## Contract with rangebar-py

rangebar-py (terrylica/opendeviationbar-py#97) will call the plugin as follows:

### Input DataFrame

```
Index: RangeIndex or DatetimeIndex
Columns guaranteed present: open, high, low, close, volume
Columns possibly present: timestamp_ms, vwap, ofi, duration_us, ... (all 55 existing columns)
Rows: sorted chronologically (ascending timestamp), single symbol, single threshold
Row count: typically 1K-500K bars per call
```

### Output DataFrame

```
Same index as input
Same existing columns unchanged (MUST NOT modify or reorder)
Additional columns: exactly those declared in self.columns
Dtypes: float64 for continuous, int64 for categorical (regime), NaN for warmup
```

### Protocol properties

| Property   | Value                                      | Purpose                                |
| ---------- | ------------------------------------------ | -------------------------------------- |
| `name`     | `"laguerre"`                               | Unique ID, used for logging/telemetry  |
| `version`  | `__version__` (e.g., `"2.1.0"`)            | Schema evolution tracking              |
| `columns`  | `("laguerre_rsi", "laguerre_regime", ...)` | rangebar-py registers these for INSERT |
| `min_bars` | `indicator.min_lookback` (typically 30-50) | rangebar-py warns if insufficient bars |

### Critical invariants

1. **Non-anticipative**: feature at bar[i] only uses data from bars[0..i]. Already guaranteed by `ATRAdaptiveLaguerreRSI` but must be preserved through the adapter.
2. **Idempotent**: `enrich(enrich(df))` == `enrich(df)`. Check if columns already exist before computing.
3. **Nullable-aware**: use `np.nan` for warmup period, NOT `0.0`. ClickHouse stores these as `Nullable(Float64) DEFAULT NULL`.
4. **No side effects**: do not modify the input DataFrame's existing columns. Only append new ones.
5. **Single config**: one `LaguerreFeatureProvider()` instance = one Laguerre config. The default is the curated Gen800 winner.

### In scope

- [ ] `LaguerreFeatureProvider` class in `src/atr_adaptive_laguerre/rangebar_plugin.py`
- [ ] Entry point registration in `pyproject.toml`
- [ ] Default config from Gen800 best performer (placeholder values initially, update when Gen800 rankings finalize)
- [ ] Environment variable override (`RANGEBAR_LAGUERRE_*`)
- [ ] Unit test: `test_rangebar_plugin.py` — verify enrich() output shape, column names, NaN warmup, non-anticipative property
- [ ] Integration test stub: verify entry point is discoverable via `importlib.metadata.entry_points`

### Out of scope

- The FeatureProvider protocol definition (lives in rangebar-py)
- ClickHouse schema migration (rangebar-py handles `ALTER TABLE ADD COLUMN`)
- Multiple simultaneous Laguerre configs (future work)
- Backfill tooling for existing ClickHouse data
- Changes to existing `fit_transform()` or `fit_transform_features()` API

## Files to Create/Modify

| File                                           | Action     | Description                                               |
| ---------------------------------------------- | ---------- | --------------------------------------------------------- |
| `src/atr_adaptive_laguerre/rangebar_plugin.py` | **CREATE** | LaguerreFeatureProvider class                             |
| `pyproject.toml`                               | **MODIFY** | Add `[project.entry-points."rangebar.feature_providers"]` |
| `tests/test_rangebar_plugin.py`                | **CREATE** | Unit tests for the plugin adapter                         |

**Zero changes** to existing files other than `pyproject.toml`.

## Why This Design

1. **Zero coupling at import time**: `rangebar_plugin.py` is only loaded when rangebar-py calls `entry_points()`. Normal `import atr_adaptive_laguerre` never touches it.
2. **rangebar-py owns the schema**: All ClickHouse schema decisions (column names, types, migration) are in rangebar-py. This package just computes values.
3. **Config is curated, not exhaustive**: 6 columns from 1 config, not 288 x 31 = 8,928 columns. More configs can be added as `laguerre2_*` if needed.
4. **Existing API unchanged**: `fit_transform()`, `fit_transform_features()`, `compute_indicator()` — all untouched. The plugin is a thin adapter layer on top.

## Downstream Dependency

This issue pairs with **terrylica/rangebar-py#98** which implements the FeatureProvider protocol, entry-point discovery, dynamic column registry, and ClickHouse INSERT integration on the rangebar-py side.

## Acceptance Criteria

1. `python -c "from atr_adaptive_laguerre.rangebar_plugin import LaguerreFeatureProvider; print('OK')"` works
2. `LaguerreFeatureProvider().columns` returns exactly `("laguerre_rsi", "laguerre_regime", "laguerre_regime_strength", "laguerre_bars_in_regime", "laguerre_tail_risk_score", "laguerre_rsi_velocity")`
3. `LaguerreFeatureProvider().enrich(ohlcv_df, "BTCUSDT", 500)` returns DataFrame with 6 additional columns
4. Warmup rows (0..min_bars-1) have NaN, remaining rows have valid float64 values
5. Entry point `rangebar.feature_providers` is discoverable: `importlib.metadata.entry_points(group="rangebar.feature_providers")` returns the `laguerre` entry
6. All existing tests pass unchanged (no regressions)
7. `enrich()` is non-anticipative (validated by existing `validate_non_anticipative()` utility)
